File: python/open_guardrail/guards/domain_allowlist.py
"""Allowlist/denylist for domains in text."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_URL_RE = re.compile(
    r"https?://([a-zA-Z0-9.-]+)"
)


def _extract_domains(text: str) -> list[str]:
    return [
        m.group(1).lower()
        for m in _URL_RE.finditer(text)
    ]


def _matches_domain(
    domain: str, pattern: str
) -> bool:
    p = pattern.lower()
    return domain == p or domain.endswith(
        "." + p
    )


class _DomainAllowlist:
    def __init__(
        self,
        *,
        action: str = "block",
        allowed_domains: Optional[List[str]] = None,
        denied_domains: Optional[List[str]] = None,
    ) -> None:
        self.name = "domain-allowlist"
        self.action = action
        self._allowed = allowed_domains or []
        self._denied = denied_domains or []

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        domains = _extract_domains(text)
        blocked: list[str] = []

        for d in domains:
            if self._denied and any(
                _matches_domain(d, p)
                for p in self._denied
            ):
                blocked.append(d)
                continue
            if self._allowed and not any(
                _matches_domain(d, p)
                for p in self._allowed
            ):
                blocked.append(d)

        triggered = len(blocked) > 0
        score = (
            min(len(blocked) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (
            time.perf_counter() - start
        ) * 1000

        return GuardResult(
            guard_name="domain-allowlist",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            score=score,
            message=(
                "Blocked domains found"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"blocked_domains": blocked}
                if triggered
                else None
            ),
        )


def domain_allowlist(
    *,
    action: str = "block",
    allowed_domains: Optional[List[str]] = None,
    denied_domains: Optional[List[str]] = None,
) -> _DomainAllowlist:
    return _DomainAllowlist(
        action=action,
        allowed_domains=allowed_domains,
        denied_domains=denied_domains,
    )

File: packages/guards/src/domain-allowlist.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface DomainAllowlistOptions {
  action: 'block' | 'warn';
  allowedDomains?: string[];
  deniedDomains?: string[];
}

const URL_RE = /https?:\/\/([a-zA-Z0-9.-]+)/g;

function extractDomains(text: string): string[] {
  const domains: string[] = [];
  let m: RegExpExecArray | null;
  while ((m = URL_RE.exec(text)) !== null) {
    domains.push(m[1].toLowerCase());
  }
  URL_RE.lastIndex = 0;
  return domains;
}

function matchesDomain(domain: string, pattern: string): boolean {
  const p = pattern.toLowerCase();
  return domain === p || domain.endsWith('.' + p);
}

export function domainAllowlist(options: DomainAllowlistOptions): Guard {
  const allowed = options.allowedDomains ?? [];
  const denied = options.deniedDomains ?? [];

  return {
    name: 'domain-allowlist',
    version: '0.1.0',
    description: 'Allowlist/denylist for domains in text',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const domains = extractDomains(text);
      const blocked: string[] = [];

      for (const d of domains) {
        if (denied.length > 0 && denied.some(p => matchesDomain(d, p))) {
          blocked.push(d);
          continue;
        }
        if (allowed.length > 0 && !allowed.some(p => matchesDomain(d, p))) {
          blocked.push(d);
        }
      }

      const triggered = blocked.length > 0;
      const score = triggered ? Math.min(blocked.length / 3, 1.0) : 0;

      return {
        guardName: 'domain-allowlist',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { blockedDomains: blocked } : undefined,
      };
    },
  };
}
